### ISSUE TITLE
Token based auth broken in 1.2.1

### ISSUE THREAD

**Issue by dhiltgen** [docker/swarm#2258]
Authentication using registry tokens was broken in 1.2.1 (and 1.2.2) - https://github.com/docker/swarm/blob/master/docs/swarm-api.md#authenticate-using-registry-tokens

The following example program can be used to test the behavior

```go
package main

import (
        "encoding/json"
        "fmt"
        "io/ioutil"
        "log"
        "net/http"
        "os"
        "path/filepath"

        "github.com/docker/docker/pkg/tlsconfig"
        "github.com/samalba/dockerclient"
)

var (
        Image = "PUT_PRIVATE_IMAGE_HERE" // TODO replace with a real private image, without tag
)

type Token struct {
        Token string `json:"token"`
}

func main() {
        fmt.Printf("Client Version: %s\n", dockerclient.APIVersion)
        certDir := os.Getenv("DOCKER_CERT_PATH")
        if certDir == "" {
                log.Fatalf("This client only works with a TLS enabled endpoint")
        }
        username := os.Getenv("REGISTRY_USERNAME")
        if username == "" {
                log.Fatalf("Set REGISTRY_USERNAME in your environment")
        }
        password := os.Getenv("REGISTRY_PASSWORD")
        if username == "" {
                log.Fatalf("Set REGISTRY_PASSWORD in your environment")
        }

        // Get the token
        client := &http.Client{}
        reqURL := fmt.Sprintf("https://auth.docker.io/token?scope=repository:%s:pull&service=registry.docker.io", Image)
        req, err := http.NewRequest("GET", reqURL, nil)
        if err != nil {
                log.Fatal(err)
        }
        req.SetBasicAuth(username, password)
        resp, err := client.Do(req)
        if err != nil {
                log.Fatal(err)
        }       
        defer resp.Body.Close()
        body, err := ioutil.ReadAll(resp.Body)
        
        var token Token
        err = json.Unmarshal(body, &token)
        if err != nil { 
                log.Fatal(err)
        }
                
        log.Printf("Got Token: %s\n", token.Token)
                
        caFilename := filepath.Join(certDir, "ca.pem")
        certFilename := filepath.Join(certDir, "cert.pem")
        keyFilename := filepath.Join(certDir, "key.pem")
        tlsConfig, err := tlsconfig.Client(tlsconfig.Options{
                CAFile:   caFilename,
                CertFile: certFilename,
                KeyFile:  keyFilename,
        })
        if err != nil {
                fmt.Printf("Failed to setup TLS: %s\n", err)
                os.Exit(1)
        }
        c, err := dockerclient.NewDockerClient(os.Getenv("DOCKER_HOST"), tlsConfig)
        if err != nil {
                fmt.Printf("Failed to connect: %s\n", err)
                os.Exit(1)
        }

        authConfig := &dockerclient.AuthConfig{RegistryToken: token.Token}

        log.Printf("Attempting to pull %s:latest\n", Image)

        err = c.PullImage(Image, authConfig)
        if err != nil {
                log.Fatalf("Failed: %s", err)
        }
        log.Println("Pull worked!")
}
```

**Comment by davgeek** [comment]
Hi @dhiltgen this script works with me, i need only run on each docker host, and then i run images from swarm.

**Comment by nishanttotla** [comment]
This was fixed in #2272
